### DISCUSSION TITLE
Freezing host application (unity freezing)

### DISCUSSION THREAD

**Discussion by devingDev** [IntergatedCircuits/HidSharp#59] [Issues]
I'm on windows using Unity 2022.3.28f1 and the current latest commit on this repo.
I tried the nuget package and also this, but neither works perfectly inside Unity.

The second time I press play it freezes up and I have to force stop it in task manager.

Found this but i don't know how to fix it right now:
https://forum.zer7.com/topic/10126/

**Comment by devingDev** [comment]
Fixed it this way if anyone cares:
In WinHidManager.cs

1. Change this part in the Run method:

```
NativeMethods.MSG msg;
while (true)
{
    int result = NativeMethods.GetMessage(out msg, hwnd, 0, 0);
    if (result == 0 || result == -1) { break; }

    NativeMethods.TranslateMessage(ref msg);
    NativeMethods.DispatchMessage(ref msg);
}
```

to:

```
NativeMethods.MSG msg;
_hwnd = hwnd;
while (true)
{
    int result = NativeMethods.GetMessage(out msg, hwnd, 0, 0);
    if (result == 0 || result == -1) { break; }

    NativeMethods.TranslateMessage(ref msg);
    NativeMethods.DispatchMessage(ref msg);
}
```


2. After or before the Run method add this:

```
		static IntPtr _hwnd;
		public static void QuitThisBs()
		{
			uint WM_QUIT = 0x0012;
			if (_hwnd != IntPtr.Zero)
			{ 
				PostMessage(_hwnd, WM_QUIT, IntPtr.Zero, IntPtr.Zero);
			}
		}

		[DllImport("user32.dll")]
		static extern bool PostMessage(IntPtr hWnd, uint Msg, IntPtr wParam, IntPtr lParam);
```




3. In your code where you exit or want to stop it call WinHidManager.QuitThisBs();
(For me I did it in the OnDestroy method Unity provides, but you can call it anywhere even prematurely while game is still running)

4. Done!

**Comment by benedekkupper** [comment]
Thanks for your report and investigation on the fix! Care to open a PR with these changes?

**Comment by Jens-roesing** [comment]
Thanks @devingDev ! 

I can confirm that this fix resolves the issues with Unity Editor on Windows (I am using Unity 2021.3.14)

However, I still had problems accessing the “QuitThisBs” function in WinHidManager from one of my own scripts, i.e. from outside. I have therefore added this to WinHidManager:

At the beginning of the Run method:
`Application.quitting += OnApplicationQuit;`

And a new method:
```
        private void OnApplicationQuit()
        {
            QuitThisBs();
        }
```

thanks again for this fix, now I don't have to restart unity every 5 seconds ;)

**Comment by PolyCrusher** [comment]
There is also a problem with the MacHidManager:

After reloading the domain/assembly, the IOHIDManager is still running and listen for device changes. 
Some cleanup is required here.

`
protected override void Run(Action readyCallback)
		{
			readyCallback();
			return;

			using (var manager = NativeMethods.IOHIDManagerCreate(IntPtr.Zero).ToCFType())
			{
				RunAssert(manager.IsSet, "HidSharp IOHIDManagerCreate failed.");

				using (var matching = NativeMethods.IOServiceMatching("IOHIDDevice").ToCFType())
				{
					RunAssert(matching.IsSet, "HidSharp IOServiceMatching failed.");

					var devicesChangedCallback = new NativeMethods.IOHIDDeviceCallback(DevicesChangedCallback);
					NativeMethods.IOHIDManagerSetDeviceMatching(manager.Handle, matching.Handle);
					NativeMethods.IOHIDManagerRegisterDeviceMatchingCallback(manager.Handle, devicesChangedCallback, IntPtr.Zero);
					NativeMethods.IOHIDManagerRegisterDeviceRemovalCallback(manager.Handle, devicesChangedCallback, IntPtr.Zero);

					var runLoop = NativeMethods.CFRunLoopGetCurrent();
					NativeMethods.CFRetain(runLoop);
					NativeMethods.IOHIDManagerScheduleWithRunLoop(manager, runLoop, NativeMethods.kCFRunLoopDefaultMode);
					try
					{
						readyCallback();
						NativeMethods.CFRunLoopRun();
					}
					finally
					{
						NativeMethods.IOHIDManagerUnscheduleFromRunLoop(manager, runLoop, NativeMethods.kCFRunLoopDefaultMode);
						NativeMethods.CFRelease(runLoop);
					}

					GC.KeepAlive(devicesChangedCallback);
				}
			}
		}
`

**Comment by OnurRafet** [comment]
> Thanks @devingDev !
> 
> I can confirm that this fix resolves the issues with Unity Editor on Windows (I am using Unity 2021.3.14)
> 
> However, I still had problems accessing the “QuitThisBs” function in WinHidManager from one of my own scripts, i.e. from outside. I have therefore added this to WinHidManager:
> 
> At the beginning of the Run method: `Application.quitting += OnApplicationQuit;`
> 
> And a new method:
> 
> ```
>         private void OnApplicationQuit()
>         {
>             QuitThisBs();
>         }
> ```
> 
> thanks again for this fix, now I don't have to restart unity every 5 seconds ;)

How did you add Application.quitting to the Run method in WinHidManager? I do not seem to have one?

**Comment by Jens-roesing** [comment]
@OnurRafet 

Just like this, right at the start a register to the quitting event, only works with in Unity because "Application" is from `using UnityEngine;`

```c#
        private void OnApplicationQuit()
        {
            QuitThisBs();
        }

        protected override void Run(Action readyCallback)
        {
            UnityEngine.Application.quitting += OnApplicationQuit;

            // Rest of the run function ...
        }
```

**Comment by viol3** [comment]
@Jens-roesing how do you use source code in unity? It gives me "use unsafe code". So I build first in cs community and I use dll only

**Comment by Jens-roesing** [comment]
> [@Jens-roesing](https://github.com/Jens-roesing) how do you use source code in unity? It gives me "use unsafe code". So I build first in cs community and I use dll only

@viol3 
Yes, I switched to a self compiled dll version, with exposed function of "QuitThisBs()". That I call inside my Project.
But you can have a look at the official website of hidsharp, the developer made an Update that sounds like a fix for this Problem:

> 2.6.3 (September 26, 2025):
Improved shutdown behavior on Windows, Mac, and Linux.
Added HidSharp.Utility.HidSharpLibrary.ManualShutdown(), for cases where the library needs to be unloaded explicitly before program shutdown.

Look at this website from the developer: https://software.seekye.com/hidsharp
